File: src/lane_detector.py
import numpy as np
import scipy as sp
import scipy.ndimage as spn
import cv2
import matplotlib.pyplot as plt

class LaneDetector:
    def __init__(self):
        self.min_pix_count = 400
        self.saliency_threshold_static = 0.400
        self.saliency_subwindows_static_x = 4
        self.saliency_subwindows_static_y = 4
        self.intensity_threshold = 100
        self.v_threshold = 0.7
        self.s_threshold = 0.1
        
    def compute_saliency(self, lab_image):
        lab_image = np.float32(lab_image) / 255.0
        m = cv2.mean(lab_image)
        saliency = lab_image[:,:,0] + lab_image[:,:,1] + lab_image[:,:,2] - m[0] - m[1] - m[2]
        b, saliency = cv2.threshold(saliency, 0.0, 1.0, 3)
        minVal, maxval, minLoc, maxLoc = cv2.minMaxLoc(saliency)
        return saliency / maxval

    def saliency_lanes_static(self, img, win_ratio_x, win_ratio_y):
        saliency_map_out = np.empty((img.shape[0], img.shape[1]))
        saliency_bin_out = np.empty((img.shape[0], img.shape[1]))
        
        nominal_win_size_x = int(round(img.shape[0] / win_ratio_x))
        nominal_win_size_y = int(round(img.shape[1] / win_ratio_y))
        
        lab_img = cv2.cvtColor(img, cv2.COLOR_BGR2LAB)
        
        for j in range(1, win_ratio_x + 1):
            for i in range(1, win_ratio_y + 1):
                top = (i-1) * nominal_win_size_x
                bottom = min(i*nominal_win_size_x, img.shape[0])
                left = (j-1) * nominal_win_size_y
                right = min(j*nominal_win_size_y, img.shape[1])
                
                subimage = lab_img[top:bottom, left:right, :]
                submap_out = self.compute_saliency(subimage)
                sub_bin_out = submap_out > self.saliency_threshold_static;
                
                saliency_map_out[top:bottom, left:right] = submap_out
                saliency_bin_out[top:bottom, left:right] = sub_bin_out
                
        return saliency_bin_out
        
    def cluster_check(self, subwindow, submap):
        verified_map_out = np.zeros((subwindow.shape[0], subwindow.shape[1]))
        temp = np.uint8(submap)
        dsubwindow = np.float64(subwindow)
        
        cluster_map, cluster_num = spn.measurements.label(temp) #default mode is 4-connected
        #counts = np.bincount(cluster_map.ravel())
        
        for i in range(1, cluster_num+1):
            if np.sum(cluster_map == i) > self.min_pix_count / 10.0:
            #if counts[i] > self.min_pix_count / 10.0:
                mean_cluster = np.mean(dsubwindow[cluster_map  == i])
                mean_else = np.mean(dsubwindow[cluster_map != i])
                
                if mean_cluster > mean_else:
                    verified_map_out[cluster_map == i] = 1
        return verified_map_out
        
    def lane_verify(self, gray_image, lane_mask, win_ratio_x, win_ratio_y):
        lane_map = np.empty(gray_image.shape).astype(np.float64)
        lane_map_out = np.empty(gray_image.shape).astype(np.float64)
        
        nominal_win_size_x = int(round(gray_image.shape[0] / win_ratio_x))
        nominal_win_size_y = int(round(gray_image.shape[1] / win_ratio_y))

        for j in range(1, win_ratio_x + 1):
            for i in range(1, win_ratio_y+1):
                top = (i-1) * nominal_win_size_x
                bottom = min(i*nominal_win_size_x, gray_image.shape[0])
                left = (j-1) * nominal_win_size_y
                right = min(j*nominal_win_size_y, gray_image.shape[1])      
                
                subwindow = gray_image[top:bottom, left:right]
                submap = lane_mask[top:bottom, left:right]
                submap_out = self.cluster_check(subwindow, submap)
                lane_map[top:bottom, left:right] = submap_out

        element = cv2.getStructuringElement(cv2.MORPH_ELLIPSE, (5,5))
        lane_map = cv2.morphologyEx(lane_map, cv2.MORPH_CLOSE, element)
        
        cluster_map, cluster_num = spn.measurements.label(np.uint8(lane_map))
        
        for i in range(1, cluster_num+1):
            area = np.sum(cluster_map == i)
            if area > self.min_pix_count:
                if np.sum(gray_image[cluster_map == i])/area > self.intensity_threshold:
                    lane_map_out[cluster_map == i] = 10
                    
        return lane_map_out
        
    def hsv_verify(self, image, v_thres, s_thres):
        hsv_image = cv2.cvtColor(image, cv2.COLOR_BGR2HSV)
        
        vMask = hsv_image[:,:,2] >= v_thres
        sMask = hsv_image[:,:,1] >= s_thres
        mask = np.logical_and(vMask, sMask)
        return mask
        
    def getLane(self, image):
        image_blurred = cv2.GaussianBlur(image, (7,7), 0)
        gray_image = cv2.cvtColor(image_blurred, cv2.COLOR_BGR2GRAY)
        
        saliency_initial_lane_map_static = self.saliency_lanes_static(image_blurred, self.saliency_subwindows_static_x, self.saliency_subwindows_static_y)
        saliency_verified_lane_map_static = self.lane_verify(gray_image, saliency_initial_lane_map_static, self.saliency_subwindows_static_x, self.saliency_subwindows_static_y)
        hsv_mask = self.hsv_verify(image, self.v_threshold, self.s_threshold)
        hsv_lane_map = hsv_mask * saliency_verified_lane_map_static
        
        #image[hsv_lane_map,0] = 0
        #image[hsv_lane_map,1] = 255
        #image[hsv_lane_map,2] = 255
        
        return hsv_lane_map
    
    
    
        
    
        
        

File: src/LaneDetector.cpp
#include "LaneDetector.h"
#include "opencv2/opencv.hpp"
#include "opencv2/core/core.hpp"
#include "opencv2/highgui/highgui.hpp"
#include <cmath>
#include "cvblob.h"

LaneDetector::LaneDetector(void)
{
    minimum_pixel_count = 400;

    saliency_threshold_static = 0.4;
	saliency_subwindows_static_x = 4;
	saliency_subwindows_static_y = 4;
	fisher_threshold_saliency_static = 0.001;

	intensity_threshold = 100;

    v_threshold = 0.7;
	s_threshold = 0.1;
}

LaneDetector::~LaneDetector(void)
{
}

inline double round( double d )
{
    return floor( d + 0.5 );
}

cv::Mat LaneDetector::compute_saliency (cv::Mat lab_image)
{
	lab_image.convertTo(lab_image, CV_32F);
	lab_image = lab_image/255;
	cv::Scalar m = mean(lab_image);
	cv::vector<cv::Mat> channels(3);
    split(lab_image, channels);
	cv::Mat saliency = channels[0] + channels[1] + channels[2] - m[0] - m[1] - m[2];
	cv::threshold(saliency, saliency, 0.0, 1.0, 3);
    double minVal;
    double maxVal;
    cv::Point minLoc;
    cv::Point maxLoc;

    minMaxLoc( saliency, &minVal, &maxVal, &minLoc, &maxLoc );

	return saliency / maxVal;
}

cv::Mat LaneDetector::saliency_lanes_static(cv::Mat image, int win_ratio_x, int win_ratio_y)
{
    cv::Size s = image.size();
    int rows = s.height;
    int cols = s.width;

    cv::Mat saliency_map_out = cv::Mat(rows, cols, CV_64F, 0.0);
    cv::Mat saliency_bin_out = cv::Mat(rows, cols, CV_64F, 0.0);

    int nominal_win_size_x = round(rows/win_ratio_x);
    int nominal_win_size_y = round(cols/win_ratio_y);

    cv::Mat lab_image;
    cvtColor( image, lab_image, CV_BGR2Lab );

    for (int j = 1; j <= win_ratio_x; j++)
        {
		for (int i = 1; i <= win_ratio_y; i++)
			{
            int top = (i - 1)*nominal_win_size_x;
            int bottom = std::min(i*nominal_win_size_x, rows);
            int left = (j - 1)*nominal_win_size_y;
            int right = std::min(j*nominal_win_size_y, cols);

            cv::Mat subimage = lab_image.colRange(left, right).rowRange(top, bottom);
            cv::Mat submap_out = compute_saliency(subimage);

            cv::Mat sub_bin_out = submap_out > saliency_threshold_static;

			submap_out = submap_out * 255;
			sub_bin_out = sub_bin_out * 255;
			submap_out.copyTo(saliency_map_out.colRange(left, right).rowRange(top, bottom));
			sub_bin_out.copyTo(saliency_bin_out.colRange(left, right).rowRange(top, bottom));
		}
    }
	return saliency_bin_out;
}

cv::Mat LaneDetector::cluster_check(cv::Mat subwindow, cv::Mat submap)
{
    cv::Size s = subwindow.size();
    int rows = s.height;
    int cols = s.width;
	
    cv::Mat verified_map_out = cv::Mat::zeros(rows, cols, CV_32F);
	cv::Mat temp;
	submap.convertTo(temp, CV_8U);
	IplImage iplsubmap = temp;
	IplImage *labelImg = cvCreateImage(s, IPL_DEPTH_LABEL, 1);
	cvb::CvBlobs blobs;
	cvb::cvLabel(&iplsubmap, labelImg, blobs);

	for (cvb::CvBlobs::const_iterator it=blobs.begin(); it!=blobs.end(); ++it)
	{
		if (it->second->area > minimum_pixel_count)
		{
			IplImage *individualBlobImg = cvCreateImage(s, 8, 3);
			cvSet(individualBlobImg, CV_RGB(0, 0, 0));
			cv::cvtColor(temp, temp, CV_GRAY2BGR);
			iplsubmap = temp;
			cvb::cvRenderBlobs(labelImg, *it->second, &iplsubmap, individualBlobImg, CV_BLOB_RENDER_COLOR);
			cv::Mat output(individualBlobImg);			
			cv::vector<cv::Mat> channels(3);
			split(output, channels);
			cv::vector<cv::Mat> channells(3);
			split(verified_map_out, channells);
			cv::Scalar mean_cluster = sum(subwindow.mul(channels[0])) / (cv::Scalar)(it->second->area);
			cv::Scalar mean_else = mean(subwindow);
			channels[0].convertTo(temp, CV_32F);
			if (mean_cluster[0] > mean_else[0])
			{
				// TO-DO: Variance stuff
				cv::bitwise_or(verified_map_out, temp, verified_map_out);
			}
		}
	}

	return verified_map_out;
}

cv::Mat LaneDetector::lane_verify(cv::Mat gray_image, cv::Mat lane_mask, int win_ratio_x, int win_ratio_y)
{
    cv::Size s = gray_image.size();
    int rows = s.height;
    int cols = s.width;

    cv::Mat lane_map = cv::Mat(rows, cols, CV_64F, 0.0);
    cv::Mat lane_map_out = cv::Mat(rows, cols, CV_64F, 0.0);
	
    int nominal_win_size_x = round(rows/win_ratio_x);
    int nominal_win_size_y = round(cols/win_ratio_y);

    for (int j = 1; j <= win_ratio_x; j++)
        {
		for (int i = 1; i <= win_ratio_y; i++)
			{
            int top = (i - 1)*nominal_win_size_x;
            int bottom = std::min(i*nominal_win_size_x, rows);
            int left = (j - 1)*nominal_win_size_y;
            int right = std::min(j*nominal_win_size_y, cols);

            cv::Mat subwindow = gray_image.colRange(left, right).rowRange(top, bottom);
            cv::Mat submap = lane_mask.colRange(left, right).rowRange(top, bottom);
            cv::Mat submap_out = cluster_check(subwindow, submap);
			submap_out.copyTo(lane_map.colRange(left, right).rowRange(top, bottom));
        }
    }

    cv::Mat element = cv::getStructuringElement(cv::MORPH_ELLIPSE, cv::Size(5, 5));
    morphologyEx(lane_map, lane_map, cv::MORPH_CLOSE, element);

	cv::Mat temp;
	lane_map.convertTo(temp, CV_8U);
	IplImage ipllanemap = temp;
	IplImage *labelImg = cvCreateImage(s, IPL_DEPTH_LABEL, 1);
	cvb::CvBlobs blobs;
	cvb::cvLabel(&ipllanemap, labelImg, blobs);
	cvb::cvFilterByArea(blobs, 5*minimum_pixel_count, 1000*minimum_pixel_count);

	IplImage *black = cvCreateImage(s, 8, 3);
	cvSet(black, CV_RGB(0, 0, 0));
	cvRenderBlobsReally(labelImg, blobs, &ipllanemap, black, CV_BLOB_RENDER_COLOR);
	cv::Mat output(black);
	cv::cvtColor(output, output, CV_BGR2GRAY, 1);

	return output;
}

cv::Mat LaneDetector::hsv_verify(cv::Mat image, int v_threshold, int s_threshold)
{
	cv::Size s = image.size();
    cv::Mat hsv_image;
    cvtColor( image, hsv_image, CV_BGR2HSV );
    cv::vector<cv::Mat> channels(3);
    split(hsv_image, channels);

	cv::Mat value_mask;
	cv::Mat saturation_mask;
	cv::threshold(channels[1], value_mask, 50, 255, 1);
	cv::threshold(channels[2], saturation_mask, 150, 255, 0);
	return value_mask.mul(saturation_mask);
}

cv::Mat LaneDetector::getLane(cv::Mat image)
{
	cv::Mat gray_image;
	cv::Mat image_blurred;
	cv::GaussianBlur(image, image_blurred, cv::Size(7, 7), 0, 0);
    cvtColor(image_blurred, gray_image, CV_BGR2GRAY );

    cv::Mat saliency_initial_lane_map_static = saliency_lanes_static(image_blurred, saliency_subwindows_static_x, saliency_subwindows_static_y);
	cv::Mat saliency_verified_lane_map_static = lane_verify(gray_image, saliency_initial_lane_map_static, saliency_subwindows_static_x, saliency_subwindows_static_y);
	
    cv::Mat hsv_mask = hsv_verify(image, v_threshold, s_threshold);
	cv::Mat hsv_lane_map;
	hsv_lane_map = saliency_verified_lane_map_static.mul(hsv_mask);

	cv::Mat yellow = cvCreateImage(image.size(), 8, 3);
	yellow = cv::Scalar(0, 255, 255);

    cv::Mat lanes_out = image;
	yellow.copyTo(lanes_out, hsv_lane_map);
	
    return lanes_out;
}

//end of LaneDetector.cpp file
